contracts/PillarStaking.sol

```diff
@@ -0,0 +1,289 @@
+        if (_stakedToken == address(0)) revert InvalidStakedToken();
+        if (_maxTotalStake > 0) maxTotalStake = _maxTotalStake;
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        stakedToken = IPillarStakingToken(_stakedToken);
+        setStateInitialized();
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 _amount) external nonReentrant whenStakeable {
+        if (block.timestamp > stakingPeriod + STAKING_PERIOD)
+            revert("StakingPeriodPassed()");
+        if ((totalStaked + _amount) > maxTotalStake)
+            revert MaximumTotalStakeReached({
+                totalMaxStake: maxTotalStake,
+                currentStakedAmount: totalStaked,
+                remainingStakeableAmount: (maxTotalStake - totalStaked),
+                stakerAmount: _amount
+            });
+        if (_amount < minStake)
+            revert InvalidMinimumStake({minimumStakeAmount: minStake});
+        if (_amount > maxStake)
+            revert InvalidMaximumStake({maximumStakeAmount: maxStake});
+        if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
+            revert StakeWouldBeGreaterThanMax();
+        if (stakingToken.balanceOf(msg.sender) < _amount)
+            revert InsufficientBalance();
+        totalStaked = totalStaked + _amount;
+        stakeholderData[msg.sender].stakedAmount =
+            stakeholderData[msg.sender].stakedAmount +
+            _amount;
+        if (!stakeholderData[msg.sender].staked) {
+            stakeholderList.push(msg.sender);
+        }
+        stakeholderData[msg.sender].staked = true;
+        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
+        stakedToken.mint(msg.sender, _amount);
+        emit Staked(msg.sender, _amount);
+    }
+
+    function unstake() external nonReentrant whenReadyForUnstake {
+        uint256 rewardAmount;
+        if (stakeholderData[msg.sender].claimed)
+            revert UserAlreadyClaimedRewards(msg.sender);
+        if (stakeholderData[msg.sender].rewardAmount == 0) {
+            calculateRewardAllocation(msg.sender);
+            rewardAmount = getRewardAmountForAccount(msg.sender);
+        } else {
+            rewardAmount = stakeholderData[msg.sender].rewardAmount;
+        }
+        uint256 stakedBalance = stakeholderData[msg.sender].stakedAmount;
+        stakeholderData[msg.sender].staked = false;
+        stakedToken.burnFrom(msg.sender, stakedBalance);
+        stakingToken.safeTransfer(msg.sender, stakedBalance);
+        stakeholderData[msg.sender].claimed = true;
+        rewardToken.safeTransfer(msg.sender, rewardAmount);
+        emit Unstaked(msg.sender, stakedBalance);
+        emit RewardPaid(msg.sender, rewardAmount);
+    }
+
+    function calculateRewardAllocation(address _staker)
+        public
+        whenReadyForUnstake
+    {
+        if (stakeholderData[msg.sender].rewardAmount != 0)
```

Comment:
Please add a check for a zero address here. Also make this internal function.

Reply:
So this was how it was originally however, Utkir raised the point that users will never be able to check their reward allocation before the rewards are transferred to them. So this should be reverted back to it original state of being baked into the unstake fn and that is it?

Reply:
I am talking about checking if the passed address parameter isn't zero. If a user wishes to check his reward balance then we shouldn't be publishing an event within this function! You can move the event to the unstake function.

Reply:
Ok, i'll put in a zero address check and remove the event. Are you happy for this to still remain a public function?